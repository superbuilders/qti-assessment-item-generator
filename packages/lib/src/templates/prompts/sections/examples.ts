import { existsSync, readFileSync } from "node:fs"
import * as path from "node:path"
import { resolveLibPath } from "@/internal/paths"

const POSITIVE_EXAMPLE_CACHE = new Map<string, string>()
const NEGATIVE_EXAMPLE_CACHE = new Map<string, string>()

const POSITIVE_ROOT = resolveLibPath("templates/prompts/examples/positive")
const POSITIVE_TEMPLATES_DIR = path.join(POSITIVE_ROOT, "templates")
const POSITIVE_NOTES_DIR = path.join(POSITIVE_ROOT, "notes")

const NEGATIVE_ROOT = resolveLibPath("templates/prompts/examples/negative")
const NEGATIVE_TEMPLATES_DIR = path.join(NEGATIVE_ROOT, "templates")
const NEGATIVE_NOTES_DIR = path.join(NEGATIVE_ROOT, "notes")

export function createPositiveExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadPositiveExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### POSITIVE_EXAMPLES
<examples type="positive">
${rendered}
</examples>`
}

export function createNegativeExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadNegativeExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### NEGATIVE_EXAMPLES
<examples type="negative">
${rendered}
</examples>`
}

function loadPositiveExample(name: string): string {
	const cached = POSITIVE_EXAMPLE_CACHE.get(name)
	if (cached) return cached

	const codePath = path.join(POSITIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}

	const notesPath = path.join(POSITIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""

	const rendered = `<example kind="positive" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	POSITIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}

function loadNegativeExample(name: string): string {
	const existing = NEGATIVE_EXAMPLE_CACHE.get(name)
	if (existing) return existing

	const codePath = path.join(NEGATIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}
	const notesPath = path.join(NEGATIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""
	const rendered = `<example kind="negative" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	NEGATIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}
