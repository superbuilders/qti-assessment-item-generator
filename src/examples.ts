import type { AssessmentItemInput } from "@/core/item"
import type { allWidgetsCollection } from "@/widgets/collections/all"
import type { WidgetTypeTupleFrom } from "@/widgets/collections/types"

export const allExamples: AssessmentItemInput<WidgetTypeTupleFrom<typeof allWidgetsCollection>>[] = []
