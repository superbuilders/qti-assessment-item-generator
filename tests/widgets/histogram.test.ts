import { describe, expect, test } from "bun:test"
import * as errors from "@superbuilders/errors"
import * as logger from "@superbuilders/slog"
import { histogramExamples } from "../../examples/histogram"
import type { WidgetInput } from "../../src/widgets/registry"
import { generateWidgetForTest } from "../helpers/generateWidgetForTest"

describe("Widget: histogram", () => {
	const examples: WidgetInput[] = histogramExamples

	examples.forEach((props, index) => {
		test(`should produce consistent output for example #${index + 1}`, async () => {
			const result = await errors.try(generateWidgetForTest(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error, index })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})
	})
})
