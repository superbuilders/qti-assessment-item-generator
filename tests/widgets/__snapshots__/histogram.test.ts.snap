// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Widget: histogram should produce consistent output for example #1 1`] = `"<svg width="692" height="517" viewBox="-73 -45 692 517" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><clipPath id="clip-0"><rect x="0" y="0" width="600" height="400"/></clipPath></defs><text x="300" y="-36.599999999999994" font-size="18" text-anchor="middle" fill="#333333" dominant-baseline="hanging"><tspan x="300" dy="0">Histogram</tspan></text><line x1="0" y1="0" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="-5" y1="400" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><text x="-8" y="404" font-size="12" text-anchor="end" fill="#333333">0</text><line x1="0" y1="400" x2="600" y2="400" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="320" x2="0" y2="320" stroke="#333333" stroke-width="1.5"/><text x="-8" y="324" font-size="12" text-anchor="end" fill="#333333">1</text><line x1="0" y1="320" x2="600" y2="320" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="240" x2="0" y2="240" stroke="#333333" stroke-width="1.5"/><text x="-8" y="244" font-size="12" text-anchor="end" fill="#333333">2</text><line x1="0" y1="240" x2="600" y2="240" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="160" x2="0" y2="160" stroke="#333333" stroke-width="1.5"/><text x="-8" y="164" font-size="12" text-anchor="end" fill="#333333">3</text><line x1="0" y1="160" x2="600" y2="160" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="80" x2="0" y2="80" stroke="#333333" stroke-width="1.5"/><text x="-8" y="84" font-size="12" text-anchor="end" fill="#333333">4</text><line x1="0" y1="80" x2="600" y2="80" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="0" x2="0" y2="0" stroke="#333333" stroke-width="1.5"/><text x="-8" y="4" font-size="12" text-anchor="end" fill="#333333">5</text><line x1="0" y1="0" x2="600" y2="0" stroke="#e0e0e0" stroke-width="1"/><text x="-54.6" y="200" font-size="16" text-anchor="middle" fill="#333333" dominant-baseline="middle" transform="rotate(-90, -54.6, 200)"><tspan x="-54.6" dy="0">Number of days</tspan></text><line x1="0" y1="400" x2="600" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="0" y1="400" x2="0" y2="405" stroke="#333333" stroke-width="1.5"/><text x="0" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">0</text><line x1="150" y1="400" x2="150" y2="405" stroke="#333333" stroke-width="1.5"/><text x="150" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">30</text><line x1="300" y1="400" x2="300" y2="405" stroke="#333333" stroke-width="1.5"/><text x="300" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">60</text><line x1="450" y1="400" x2="450" y2="405" stroke="#333333" stroke-width="1.5"/><text x="450" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">90</text><line x1="600" y1="400" x2="600" y2="405" stroke="#333333" stroke-width="1.5"/><text x="600" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">120</text><text x="300" y="447" font-size="16" text-anchor="middle" fill="#333333"><tspan x="300" dy="0">Number of emails sent</tspan></text><g clip-path="url(#clip-0)"><rect x="0" y="320" width="150" height="80" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="150" y="240" width="150" height="160" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="300" y="240" width="150" height="160" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="450" y="240" width="150" height="160" fill="#4472c4" stroke="#333333" stroke-width="1"/></g></svg>"`;

exports[`Widget: histogram should produce consistent output for example #2 1`] = `"<svg width="692" height="517" viewBox="-73 -45 692 517" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><clipPath id="clip-0"><rect x="0" y="0" width="600" height="400"/></clipPath></defs><text x="300" y="-36.599999999999994" font-size="18" text-anchor="middle" fill="#333333" dominant-baseline="hanging"><tspan x="300" dy="0">Histogram</tspan></text><line x1="0" y1="0" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="-5" y1="400" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><text x="-8" y="404" font-size="12" text-anchor="end" fill="#333333">0</text><line x1="0" y1="400" x2="600" y2="400" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="320" x2="0" y2="320" stroke="#333333" stroke-width="1.5"/><text x="-8" y="324" font-size="12" text-anchor="end" fill="#333333">1</text><line x1="0" y1="320" x2="600" y2="320" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="240" x2="0" y2="240" stroke="#333333" stroke-width="1.5"/><text x="-8" y="244" font-size="12" text-anchor="end" fill="#333333">2</text><line x1="0" y1="240" x2="600" y2="240" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="160" x2="0" y2="160" stroke="#333333" stroke-width="1.5"/><text x="-8" y="164" font-size="12" text-anchor="end" fill="#333333">3</text><line x1="0" y1="160" x2="600" y2="160" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="80" x2="0" y2="80" stroke="#333333" stroke-width="1.5"/><text x="-8" y="84" font-size="12" text-anchor="end" fill="#333333">4</text><line x1="0" y1="80" x2="600" y2="80" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="0" x2="0" y2="0" stroke="#333333" stroke-width="1.5"/><text x="-8" y="4" font-size="12" text-anchor="end" fill="#333333">5</text><line x1="0" y1="0" x2="600" y2="0" stroke="#e0e0e0" stroke-width="1"/><text x="-54.6" y="200" font-size="16" text-anchor="middle" fill="#333333" dominant-baseline="middle" transform="rotate(-90, -54.6, 200)"><tspan x="-54.6" dy="0">Number of days</tspan></text><line x1="0" y1="400" x2="600" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="0" y1="400" x2="0" y2="405" stroke="#333333" stroke-width="1.5"/><text x="0" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">0</text><line x1="150" y1="400" x2="150" y2="405" stroke="#333333" stroke-width="1.5"/><text x="150" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">30</text><line x1="300" y1="400" x2="300" y2="405" stroke="#333333" stroke-width="1.5"/><text x="300" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">60</text><line x1="450" y1="400" x2="450" y2="405" stroke="#333333" stroke-width="1.5"/><text x="450" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">90</text><line x1="600" y1="400" x2="600" y2="405" stroke="#333333" stroke-width="1.5"/><text x="600" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">120</text><text x="300" y="447" font-size="16" text-anchor="middle" fill="#333333"><tspan x="300" dy="0">Number of emails sent</tspan></text><g clip-path="url(#clip-0)"><rect x="0" y="320" width="150" height="80" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="150" y="320" width="150" height="80" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="300" y="160" width="150" height="240" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="450" y="240" width="150" height="160" fill="#4472c4" stroke="#333333" stroke-width="1"/></g></svg>"`;

exports[`Widget: histogram should produce consistent output for example #3 1`] = `"<svg width="692" height="517" viewBox="-73 -45 692 517" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><clipPath id="clip-0"><rect x="0" y="0" width="600" height="400"/></clipPath></defs><text x="300" y="-36.599999999999994" font-size="18" text-anchor="middle" fill="#333333" dominant-baseline="hanging"><tspan x="300" dy="0">Histogram</tspan></text><line x1="0" y1="0" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="-5" y1="400" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><text x="-8" y="404" font-size="12" text-anchor="end" fill="#333333">0</text><line x1="0" y1="400" x2="600" y2="400" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="320" x2="0" y2="320" stroke="#333333" stroke-width="1.5"/><text x="-8" y="324" font-size="12" text-anchor="end" fill="#333333">1</text><line x1="0" y1="320" x2="600" y2="320" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="240" x2="0" y2="240" stroke="#333333" stroke-width="1.5"/><text x="-8" y="244" font-size="12" text-anchor="end" fill="#333333">2</text><line x1="0" y1="240" x2="600" y2="240" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="160" x2="0" y2="160" stroke="#333333" stroke-width="1.5"/><text x="-8" y="164" font-size="12" text-anchor="end" fill="#333333">3</text><line x1="0" y1="160" x2="600" y2="160" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="80" x2="0" y2="80" stroke="#333333" stroke-width="1.5"/><text x="-8" y="84" font-size="12" text-anchor="end" fill="#333333">4</text><line x1="0" y1="80" x2="600" y2="80" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="0" x2="0" y2="0" stroke="#333333" stroke-width="1.5"/><text x="-8" y="4" font-size="12" text-anchor="end" fill="#333333">5</text><line x1="0" y1="0" x2="600" y2="0" stroke="#e0e0e0" stroke-width="1"/><text x="-54.6" y="200" font-size="16" text-anchor="middle" fill="#333333" dominant-baseline="middle" transform="rotate(-90, -54.6, 200)"><tspan x="-54.6" dy="0">Number of days</tspan></text><line x1="0" y1="400" x2="600" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="0" y1="400" x2="0" y2="405" stroke="#333333" stroke-width="1.5"/><text x="0" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">0</text><line x1="150" y1="400" x2="150" y2="405" stroke="#333333" stroke-width="1.5"/><text x="150" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">30</text><line x1="300" y1="400" x2="300" y2="405" stroke="#333333" stroke-width="1.5"/><text x="300" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">60</text><line x1="450" y1="400" x2="450" y2="405" stroke="#333333" stroke-width="1.5"/><text x="450" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">90</text><line x1="600" y1="400" x2="600" y2="405" stroke="#333333" stroke-width="1.5"/><text x="600" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">120</text><text x="300" y="447" font-size="16" text-anchor="middle" fill="#333333"><tspan x="300" dy="0">Number of emails sent</tspan></text><g clip-path="url(#clip-0)"><rect x="0" y="400" width="150" height="0" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="150" y="160" width="150" height="240" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="300" y="240" width="150" height="160" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="450" y="240" width="150" height="160" fill="#4472c4" stroke="#333333" stroke-width="1"/></g></svg>"`;

exports[`Widget: histogram should produce consistent output for example #4 1`] = `"<svg width="688" height="517" viewBox="-73 -45 688 517" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><clipPath id="clip-0"><rect x="0" y="0" width="600" height="400"/></clipPath></defs><text x="300" y="-36.599999999999994" font-size="18" text-anchor="middle" fill="#333333" dominant-baseline="hanging"><tspan x="300" dy="0">Red Candy Pieces per Bag</tspan></text><line x1="0" y1="0" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="-5" y1="400" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><text x="-8" y="404" font-size="12" text-anchor="end" fill="#333333">0</text><line x1="0" y1="400" x2="600" y2="400" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="320" x2="0" y2="320" stroke="#333333" stroke-width="1.5"/><text x="-8" y="324" font-size="12" text-anchor="end" fill="#333333">1</text><line x1="0" y1="320" x2="600" y2="320" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="240" x2="0" y2="240" stroke="#333333" stroke-width="1.5"/><text x="-8" y="244" font-size="12" text-anchor="end" fill="#333333">2</text><line x1="0" y1="240" x2="600" y2="240" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="160" x2="0" y2="160" stroke="#333333" stroke-width="1.5"/><text x="-8" y="164" font-size="12" text-anchor="end" fill="#333333">3</text><line x1="0" y1="160" x2="600" y2="160" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="80" x2="0" y2="80" stroke="#333333" stroke-width="1.5"/><text x="-8" y="84" font-size="12" text-anchor="end" fill="#333333">4</text><line x1="0" y1="80" x2="600" y2="80" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="0" x2="0" y2="0" stroke="#333333" stroke-width="1.5"/><text x="-8" y="4" font-size="12" text-anchor="end" fill="#333333">5</text><line x1="0" y1="0" x2="600" y2="0" stroke="#e0e0e0" stroke-width="1"/><text x="-54.6" y="200" font-size="16" text-anchor="middle" fill="#333333" dominant-baseline="middle" transform="rotate(-90, -54.6, 200)"><tspan x="-54.6" dy="0">Number of bags</tspan></text><line x1="0" y1="400" x2="600" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="0" y1="400" x2="0" y2="405" stroke="#333333" stroke-width="1.5"/><text x="0" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">0</text><line x1="150" y1="400" x2="150" y2="405" stroke="#333333" stroke-width="1.5"/><text x="150" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">3</text><line x1="300" y1="400" x2="300" y2="405" stroke="#333333" stroke-width="1.5"/><text x="300" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">6</text><line x1="450" y1="400" x2="450" y2="405" stroke="#333333" stroke-width="1.5"/><text x="450" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">9</text><line x1="600" y1="400" x2="600" y2="405" stroke="#333333" stroke-width="1.5"/><text x="600" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">12</text><text x="300" y="447" font-size="16" text-anchor="middle" fill="#333333"><tspan x="300" dy="0">Number of red candy pieces</tspan></text><g clip-path="url(#clip-0)"><rect x="0" y="320" width="150" height="80" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="150" y="240" width="150" height="160" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="300" y="240" width="150" height="160" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="450" y="240" width="150" height="160" fill="#4472c4" stroke="#333333" stroke-width="1"/></g></svg>"`;

exports[`Widget: histogram should produce consistent output for example #5 1`] = `"<svg width="688" height="517" viewBox="-73 -45 688 517" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><clipPath id="clip-0"><rect x="0" y="0" width="600" height="400"/></clipPath></defs><text x="300" y="-36.599999999999994" font-size="18" text-anchor="middle" fill="#333333" dominant-baseline="hanging"><tspan x="300" dy="0">Red Candy Pieces per Bag</tspan></text><line x1="0" y1="0" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="-5" y1="400" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><text x="-8" y="404" font-size="12" text-anchor="end" fill="#333333">0</text><line x1="0" y1="400" x2="600" y2="400" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="320" x2="0" y2="320" stroke="#333333" stroke-width="1.5"/><text x="-8" y="324" font-size="12" text-anchor="end" fill="#333333">1</text><line x1="0" y1="320" x2="600" y2="320" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="240" x2="0" y2="240" stroke="#333333" stroke-width="1.5"/><text x="-8" y="244" font-size="12" text-anchor="end" fill="#333333">2</text><line x1="0" y1="240" x2="600" y2="240" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="160" x2="0" y2="160" stroke="#333333" stroke-width="1.5"/><text x="-8" y="164" font-size="12" text-anchor="end" fill="#333333">3</text><line x1="0" y1="160" x2="600" y2="160" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="80" x2="0" y2="80" stroke="#333333" stroke-width="1.5"/><text x="-8" y="84" font-size="12" text-anchor="end" fill="#333333">4</text><line x1="0" y1="80" x2="600" y2="80" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="0" x2="0" y2="0" stroke="#333333" stroke-width="1.5"/><text x="-8" y="4" font-size="12" text-anchor="end" fill="#333333">5</text><line x1="0" y1="0" x2="600" y2="0" stroke="#e0e0e0" stroke-width="1"/><text x="-54.6" y="200" font-size="16" text-anchor="middle" fill="#333333" dominant-baseline="middle" transform="rotate(-90, -54.6, 200)"><tspan x="-54.6" dy="0">Number of bags</tspan></text><line x1="0" y1="400" x2="600" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="0" y1="400" x2="0" y2="405" stroke="#333333" stroke-width="1.5"/><text x="0" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">0</text><line x1="150" y1="400" x2="150" y2="405" stroke="#333333" stroke-width="1.5"/><text x="150" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">3</text><line x1="300" y1="400" x2="300" y2="405" stroke="#333333" stroke-width="1.5"/><text x="300" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">6</text><line x1="450" y1="400" x2="450" y2="405" stroke="#333333" stroke-width="1.5"/><text x="450" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">9</text><line x1="600" y1="400" x2="600" y2="405" stroke="#333333" stroke-width="1.5"/><text x="600" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">12</text><text x="300" y="447" font-size="16" text-anchor="middle" fill="#333333"><tspan x="300" dy="0">Number of red candy pieces</tspan></text><g clip-path="url(#clip-0)"><rect x="0" y="320" width="150" height="80" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="150" y="320" width="150" height="80" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="300" y="240" width="150" height="160" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="450" y="160" width="150" height="240" fill="#4472c4" stroke="#333333" stroke-width="1"/></g></svg>"`;

exports[`Widget: histogram should produce consistent output for example #6 1`] = `"<svg width="688" height="517" viewBox="-73 -45 688 517" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><clipPath id="clip-0"><rect x="0" y="0" width="600" height="400"/></clipPath></defs><text x="300" y="-36.599999999999994" font-size="18" text-anchor="middle" fill="#333333" dominant-baseline="hanging"><tspan x="300" dy="0">Red Candy Pieces per Bag</tspan></text><line x1="0" y1="0" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="-5" y1="400" x2="0" y2="400" stroke="#333333" stroke-width="1.5"/><text x="-8" y="404" font-size="12" text-anchor="end" fill="#333333">0</text><line x1="0" y1="400" x2="600" y2="400" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="320" x2="0" y2="320" stroke="#333333" stroke-width="1.5"/><text x="-8" y="324" font-size="12" text-anchor="end" fill="#333333">1</text><line x1="0" y1="320" x2="600" y2="320" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="240" x2="0" y2="240" stroke="#333333" stroke-width="1.5"/><text x="-8" y="244" font-size="12" text-anchor="end" fill="#333333">2</text><line x1="0" y1="240" x2="600" y2="240" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="160" x2="0" y2="160" stroke="#333333" stroke-width="1.5"/><text x="-8" y="164" font-size="12" text-anchor="end" fill="#333333">3</text><line x1="0" y1="160" x2="600" y2="160" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="80" x2="0" y2="80" stroke="#333333" stroke-width="1.5"/><text x="-8" y="84" font-size="12" text-anchor="end" fill="#333333">4</text><line x1="0" y1="80" x2="600" y2="80" stroke="#e0e0e0" stroke-width="1"/><line x1="-5" y1="0" x2="0" y2="0" stroke="#333333" stroke-width="1.5"/><text x="-8" y="4" font-size="12" text-anchor="end" fill="#333333">5</text><line x1="0" y1="0" x2="600" y2="0" stroke="#e0e0e0" stroke-width="1"/><text x="-54.6" y="200" font-size="16" text-anchor="middle" fill="#333333" dominant-baseline="middle" transform="rotate(-90, -54.6, 200)"><tspan x="-54.6" dy="0">Number of bags</tspan></text><line x1="0" y1="400" x2="600" y2="400" stroke="#333333" stroke-width="1.5"/><line x1="0" y1="400" x2="0" y2="405" stroke="#333333" stroke-width="1.5"/><text x="0" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">0</text><line x1="150" y1="400" x2="150" y2="405" stroke="#333333" stroke-width="1.5"/><text x="150" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">3</text><line x1="300" y1="400" x2="300" y2="405" stroke="#333333" stroke-width="1.5"/><text x="300" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">6</text><line x1="450" y1="400" x2="450" y2="405" stroke="#333333" stroke-width="1.5"/><text x="450" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">9</text><line x1="600" y1="400" x2="600" y2="405" stroke="#333333" stroke-width="1.5"/><text x="600" y="413" font-size="12" text-anchor="middle" dominant-baseline="hanging" fill="#333333">12</text><text x="300" y="447" font-size="16" text-anchor="middle" fill="#333333"><tspan x="300" dy="0">Number of red candy pieces</tspan></text><g clip-path="url(#clip-0)"><rect x="0" y="240" width="150" height="160" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="150" y="320" width="150" height="80" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="300" y="320" width="150" height="80" fill="#4472c4" stroke="#333333" stroke-width="1"/><rect x="450" y="160" width="150" height="240" fill="#4472c4" stroke="#333333" stroke-width="1"/></g></svg>"`;
